WordPress/Classes/ViewRelated/Notifications/Controllers/NotificationsViewController+Extensions.swift

```diff
@@ -241,6 +246,128 @@ extension NotificationsViewController: WPTableViewHandlerDelegate
 
 
 
+// MARK - Filter Helpers
+//
+extension NotificationsViewController
+{
+    func showFiltersSegmentedControlIfApplicable() {
+        guard tableHeaderView.alpha == WPAlphaZero && shouldDisplayFilters == true else {
+            return
+        }
+
+        UIView.animateWithDuration(WPAnimationDurationDefault) {
+            self.tableHeaderView.alpha = WPAlphaFull
+        }
+    }
+
+    func hideFiltersSegmentedControlIfApplicable() {
+        if tableHeaderView.alpha == WPAlphaFull && shouldDisplayFilters == false {
+            tableHeaderView.alpha = WPAlphaZero
+        }
+    }
+
+    var shouldDisplayFilters: Bool {
+        // Note:
+        // Filters should only be hidden whenever there are no Notifications in the bucket (contrary to the FRC's
+        // results, which are filtered by the active predicate!).
+        //
+        let bucket = simperium.bucketForName(entityName())
+        return bucket.numObjects() > 0
+    }
+}
+
+
+
+// MARK: - NoResults Helpers
+//
+extension NotificationsViewController
+{
+    func showNoResultsViewIfNeeded() {
+        // Remove + Show Filters, if needed
+        guard shouldDisplayNoResultsView == true else {
+            noResultsView.removeFromSuperview()
+            showFiltersSegmentedControlIfApplicable()
+            return
+        }
+
+        // Attach the view
+        if noResultsView.superview == nil {
+            tableView.addSubviewWithFadeAnimation(noResultsView)
+        }
+
+        // Refresh its properties: The user may have signed into WordPress.com
+        noResultsView.titleText     = noResultsTitleText
+        noResultsView.titleText     = noResultsTitleText
+        noResultsView.messageText   = noResultsMessageText
+        noResultsView.accessoryView = noResultsAccessoryView
+        noResultsView.buttonTitle   = noResultsButtonText
+
+        // Hide the filter header if we're showing the Jetpack prompt
+        hideFiltersSegmentedControlIfApplicable()
+    }
+
+    var noResultsTitleText: String {
+        guard shouldDisplayJetpackMessage == false else {
+            return NSLocalizedString("Connect to Jetpack", comment: "Notifications title displayed when a self-hosted user is not connected to Jetpack")
+        }
+
+        let messageMap: [Filter: String] = [
```

Comment:
How about Swifting this up a notch? You can simplify things a bit with a `switch`:

```
let filter = Filter(rawValue: filtersSegmentedControl.selectedSegmentIndex) ?? .None
switch filter {
case .None:
    return NSLocalizedString("No notifications yet", comment: "Displayed in the Notifications Tab, when there are no notifications")
case .Unread:
    return NSLocalizedString("No unread notifications", comment: "Displayed in the Notifications Tab, when the Unread Filter shows no notifications")
case .Comment:
    return NSLocalizedString("No comments notifications", comment: "Displayed in the Notifications Tab, when the Comments Filter shows no notifications")
case .Follow:
    return NSLocalizedString("No new followers notifications", comment: "Displayed in the Notifications Tab, when the Follow Filter shows no notifications")
case .Like:
    return NSLocalizedString("No like notifications", comment: "Displayed in the Notifications Tab, when the Likes Filter shows no notifications")
}
```